Svc/BufferManager/BufferManagerComponentImpl.cpp

```diff
@@ -0,0 +1,223 @@
+  // ----------------------------------------------------------------------
+
+  void BufferManagerComponentImpl ::
+    bufferSendIn_handler(
+        const NATIVE_INT_TYPE portNum,
+        Fw::Buffer &fwBuffer
+    )
+  {
+      // make sure component has been set up
+      FW_ASSERT(this->m_setup);
+      // check for empty buffers - this is just a warning since this component returns
+      // empty buffers if it can't allocate one.
+      if (fwBuffer.getSize() == 0) {
+          this->log_WARNING_HI_ZeroSizeBuffer();
+          this->m_emptyBuffs++;
+          return;
+      }
+      // use the bufferID member field to find the original slot
+      U32 context = fwBuffer.getContext();
+      U32 id = context & 0xFFFF;
+      U32 mgrId = context >> 16;
+      // check some things
+      FW_ASSERT(id < this->m_numStructs,id,this->m_numStructs);
+      FW_ASSERT(mgrId == this->m_mgrId,mgrId,id,this->m_mgrId);
+      FW_ASSERT(true == this->m_buffers[id].allocated,id,this->m_mgrId);
+      FW_ASSERT(reinterpret_cast<U8*>(fwBuffer.getData()) >= this->m_buffers[id].memory,id,this->m_mgrId);
+      FW_ASSERT(reinterpret_cast<U8*>(fwBuffer.getData()) < (this->m_buffers[id].memory + this->m_buffers[id].size),id,this->m_mgrId);
+      // user can make smaller for their own purposes, but it shouldn't be bigger
+      FW_ASSERT(fwBuffer.getSize() <= this->m_buffers[id].size,id,this->m_mgrId);
+      // clear the allocated flag
+      this->m_buffers[id].allocated = false;
+      this->m_currBuffs--;
+  }
+
+  Fw::Buffer BufferManagerComponentImpl ::
+    bufferGetCallee_handler(
+        const NATIVE_INT_TYPE portNum,
+        U32 size
+    )
+  {
+      // make sure component has been set up
+      FW_ASSERT(this->m_setup);
+      // find smallest buffer based on size.
+      for (NATIVE_UINT_TYPE buff = 0; buff < this->m_numStructs; buff++) {
+          if ((not this->m_buffers[buff].allocated) and (size < this->m_buffers[buff].size)) {
+              this->m_buffers[buff].allocated = true;
+              this->m_currBuffs++;
+              if (this->m_currBuffs > this->m_highWater) {
+                  this->m_highWater = this->m_currBuffs;
+              }
+              Fw::Buffer copy = this->m_buffers[buff].buff;
+              // change size to match request
+              copy.setSize(size);
+              return copy;
+          }
+      }
+
+      // if no buffers found, return empty buffer
+      this->log_WARNING_HI_NoBuffsAvailable(size);
+      this->m_noBuffs++;
+      return Fw::Buffer();
+
+  }
+
+  void BufferManagerComponentImpl::setup(
+    NATIVE_UINT_TYPE mgrId, //!< manager ID
+    NATIVE_UINT_TYPE memId, //!< Memory segment identifier
+    Fw::MemAllocator& allocator, //!< memory allocator
+    const BufferBins& bins //!< Set of user bins
+  ) {
+
+    this->m_mgrId = mgrId;
+    this->m_memId = memId;
+    this->m_allocator = &allocator;
+    // clear bins
+    memset(&this->m_bufferBins,0,sizeof(this->m_bufferBins));
```

Comment:
Not sure if needed, but might be safer to ASSERT NULL `this->bufferBins` before passing it to memset